src/basic/slow-code.h

```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#pragma once
+
+#include "log.h"
+#include "string-util.h"
+#include "time-util.h"
+
+#include "macro.h"
+
+typedef struct slow_code_detector {
+        uint32_t timeout_sec;
+        uint32_t count;
+        usec_t current_time;
+        size_t line_num;
+        char *file;
+        char *function;
+} slow_code_detector;
+
+static inline void slow_code(slow_code_detector *d) {
+        if (d->current_time != USEC_INFINITY && d->current_time + d->timeout_sec * USEC_PER_SEC < now(CLOCK_MONOTONIC)) {
+                log_debug("Detected overly slow code block in function %s() of %s: %" PRIu64 ".", d->function, d->file, d->line_num);
+        }
+        free(d->file);
+        free(d->function);
+}
+
+#define PROFILE_SLOW_CODE(timeout_sec) \
+        _cleanup_(slow_code) slow_code_detector detector_tmp = ((const slow_code_detector) {timeout_sec, 0,            \
+                                                                 DEBUG_LOGGING ? now(CLOCK_MONOTONIC) : USEC_INFINITY, \
+                                                                 __LINE__, strdup(__FILE__), strdup(__FUNCTION__)})
```

Comment:
no need for strdup(). the `__FILE__` and `__FUNCTION__` macros return static constant buffers that stay valid for good.

Reply:
please line break this as we usually would.

also drop the `(const slow_code_detector)`, it's not needed. It turns a regular struct initialization into a copy from a compound initialized variable, which is not desirable (hopefully the compiler will suppress the difference, but still, let's not waste complexity on that.

please always specify the field names you are initializing.

please always keep all time values in µs (and hence the usec_t type). Across our codebase we uniformly calculate in µs, please stick to that here.

please always enclose all macro params in `()` when using them (to avoid any doubts how C could interpret this in its use)

i.e. something like this:

```c
#define PROFILE_SLOW_CODE(timeout_usec) \
        _cleanup_(slow_code) slow_code_detector detector_tmp = { \
                .timeout_usec = DEBUG_LOGGING ? usec_add(now(CLOCK_MONOTONIC), (timeout_usec)) : USEC_INFINITY, \
                .line = __LINE__, \
                .file = __FILE__, \
                .func = __func__, \
        }
```